docs/staking.md

```diff
@@ -0,0 +1,258 @@
+  - **Invariant**: Decremented before creating a withdrawal.
+
+- `totalStakerStakes[staker] -> uint256`
+  - **Purpose**: Total active stake per staker across validators.
+
+- `withdrawalQueues[staker][validator] -> WithdrawalQueue`
+  - **Purpose**: Head/tail pointers for each queue.
+
+- `withdrawalNodes[withdrawalId] -> WithdrawalNode`
+  - **Purpose**: Storage of all withdrawal nodes.
+  - **Invariant**: A node exists iff `claimableAt != 0` and referenced by a queue.
+
+- `pendingWithdrawDelayChange -> ConfigProposal`
+  - **Purpose**: Pending withdraw delay change.
+
+- `pendingValidatorChangeHash -> bytes32`
+  - **Purpose**: Hash of a pending validator update proposal.
+  - **Edge case**: Only the hash is stored; execution requires supplying matching arrays.
+
+## Events
+- `StakeIncreased`: Emitted after stake state is updated and before token transfer.
+- `WithdrawalInitiated`: Emitted when a withdrawal node is created.
+- `WithdrawalClaimed`: Emitted after queue updates and before token transfer.
+- `ValidatorsProposed`: Emitted for validator update proposals; includes arrays and `executableAt`.
+- `ValidatorUpdated`: Emitted per validator when a proposal is executed.
+- `WithdrawDelayProposed`: Emitted on withdraw delay proposal.
+- `WithdrawDelayChanged`: Emitted on execution of withdraw delay change (also emitted in constructor).
+- `TokensRecovered`: Emitted before token recovery transfer.
+
+**Event strategy**: Every state-changing external action emits a dedicated event to support off-chain indexing and auditability.
+
+## Errors
+Custom errors are used for gas efficiency and precise revert reasons:
+
+- `InvalidAmount`: Zero or invalid amount.
+- `InvalidAddress`: Zero address input.
+- `NotValidator`: Staking to non-registered validator.
+- `InsufficientStake`: Withdraw amount exceeds current stake.
+- `ProposalNotSet`, `InvalidProposalHash`, `ProposalNotExecutable`, `NoProposalExists`: Timelock and proposal validation.
+- `InsufficientRecoverableAmount`: No recoverable tokens.
+- `WithdrawalQueueEmpty`, `NoClaimableWithdrawal`: Withdrawal claim validation.
+- `ArrayLengthMismatch`: Validators arrays mismatch.
+- `InvalidParameter`: Parameter out of bounds (e.g., withdraw delay).
+- `InvalidWithdrawalNode`, `InvalidOrdering`: Manual queue insertion validation.
+
+## External Functions
+
+### Staking Operations
+- `stake(validator, amount)`
+  - **Purpose**: Increase stake for a validator.
+  - **Invariants**: Validator must be registered; amount must be non-zero.
+  - **Security**: State updates precede `safeTransferFrom` (reverts roll back state).
+
+- `initiateWithdrawal(validator, amount)`
+  - **Purpose**: Create a withdrawal and insert into a sorted queue automatically.
+  - **Edge case**: Traverses from tail; long queues may be gas-intensive.
+  - **Ordering**: Ensures queue remains sorted by `claimableAt`.
+
+- `initiateWithdrawalAtPosition(validator, amount, previousId)`
```

Comment:
Might we worth expanding on the purpose here (it was because we wanted to maintain a sorted list, and to deal with edge cases around changes to the delay). AFAIU, `initiateWithdrawal` should be used 99.99% of the time.